### PR TITLE
[TE-508] Read and use Capabilities and Options

### PR DESCRIPTION
**Pull request by kbayraktar**

### CODE REVIEW

File: src/main/java/org/testeditor/fixture/web/WebDriverFixture.java [hunk 1]
```diff
@@ -128,7 +138,8 @@ public void setExecuteScript(String exec) {
      * <li><b>chrome</b> - opens Google Chrome</li>
      * </ul>
      * 
-     * @param browser String literal for used browser
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/fixture/web/WebDriverFixture.java]: please put those changes in a pretty print pull in the future, it distracts from reviewing this pull.

File: src/main/java/org/testeditor/fixture/web/WebDriverFixture.java [hunk 2]
```diff
@@ -266,7 +278,11 @@ public WebDriver startFireFoxPortable(String browserPath) {
      */
     private void launchChrome() {
         setupDrivermanager(ChromeDriverManager.getInstance());
-        driver = new ChromeDriver();
+        String browserName = BrowserType.CHROME;
+        final DesiredCapabilities desiredCapabilities = DesiredCapabilities.chrome();
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/fixture/web/WebDriverFixture.java]: final is cool, but why are then the other field not final?

File: src/main/java/org/testeditor/fixture/web/WebDriverFixture.java [hunk 3]
```diff
@@ -277,7 +293,11 @@ private void launchChrome() {
      */
     private void launchFirefox() {
         setupDrivermanager(FirefoxDriverManager.getInstance());
-        driver = new FirefoxDriver();
+        String browserName = BrowserType.FIREFOX;
+        final DesiredCapabilities desiredCapabilities = DesiredCapabilities.firefox();
+        Object firefoxOptions = populateBrowserSettings(desiredCapabilities, browserName);
+        desiredCapabilities.setCapability(FirefoxOptions.FIREFOX_OPTIONS, firefoxOptions);
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/fixture/web/WebDriverFixture.java]: setting capabilities with options? there seems to be a naming problem.

File: src/main/java/org/testeditor/fixture/web/WebDriverFixture.java [hunk 4]
```diff
@@ -288,15 +308,115 @@ private void launchFirefox() {
...
+        DesiredCapabilities desiredCapabilities = DesiredCapabilities.internetExplorer();
+        setCapabilitiesForIe(desiredCapabilities, browserName);
+        driver = new InternetExplorerDriver(desiredCapabilities);
         registerShutdownHook(driver);
     }
 
-    private void setCapabilities(DesiredCapabilities capabilities) {
-        capabilities.setCapability(InternetExplorerDriver.INTRODUCE_FLAKINESS_BY_IGNORING_SECURITY_DOMAINS, true);
-        capabilities.setCapability("requireWindowFocus", true);
+    private void setCapabilitiesForIe(DesiredCapabilities desiredCapabilities, String browserName) {
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/fixture/web/WebDriverFixture.java]: why is this ie specific? could this be generic?

File: src/main/java/org/testeditor/fixture/web/WebDriverFixture.java [hunk 5]
```diff
@@ -277,7 +293,11 @@ private void launchChrome() {
      */
     private void launchFirefox() {
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/fixture/web/WebDriverFixture.java]: these methods are all very similar, could they be made generic with BrowserType as parameter?

File: src/main/java/org/testeditor/fixture/web/WebDriverFixture.java [hunk 6]
```diff
@@ -288,15 +308,115 @@ private void launchFirefox() {
...
-        capabilities.setCapability(InternetExplorerDriver.INTRODUCE_FLAKINESS_BY_IGNORING_SECURITY_DOMAINS, true);
-        capabilities.setCapability("requireWindowFocus", true);
+    private void setCapabilitiesForIe(DesiredCapabilities desiredCapabilities, String browserName) {
+        List<BrowserSetting> capabilities = new ArrayList<>();
+        List<BrowserSetting> options = new ArrayList<>();
+        populateBrowserSpecificSettings(browserName, capabilities, options);
+        populateCapabilities(desiredCapabilities, capabilities);
+    }
+
+    private Object populateBrowserSettings(DesiredCapabilities desiredCapabilities, String browserName) {
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/fixture/web/WebDriverFixture.java]: BrowserType could be an enum, what do you think?

File: src/main/java/org/testeditor/fixture/web/WebDriverFixture.java [hunk 7]
```diff
@@ -288,15 +308,115 @@ private void launchFirefox() {
...
+        List<BrowserSetting> capabilities = new ArrayList<>();
+        List<BrowserSetting> options = new ArrayList<>();
+        // über den Browser die cap und options holen
+        populateBrowserSpecificSettings(browserName, capabilities, options);
+        // set all capabilities Capabilities
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/fixture/web/WebDriverFixture.java]: what do you mean by this comment?

File: src/main/java/org/testeditor/fixture/web/WebDriverFixture.java [hunk 8]
```diff
@@ -288,15 +308,115 @@ private void launchFirefox() {
...
+        List<BrowserSetting> capabilities = new ArrayList<>();
+        List<BrowserSetting> options = new ArrayList<>();
+        populateBrowserSpecificSettings(browserName, capabilities, options);
+        populateCapabilities(desiredCapabilities, capabilities);
+    }
+
+    private Object populateBrowserSettings(DesiredCapabilities desiredCapabilities, String browserName) {
+        List<BrowserSetting> capabilities = new ArrayList<>();
+        List<BrowserSetting> options = new ArrayList<>();
+        // über den Browser die cap und options holen
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/fixture/web/WebDriverFixture.java]: please use english comments

File: src/main/java/org/testeditor/fixture/web/WebDriverFixture.java [hunk 9]
```diff
@@ -288,15 +308,115 @@ private void launchFirefox() {
...
+        populateBrowserSpecificSettings(browserName, capabilities, options);
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/fixture/web/WebDriverFixture.java]: I think a better naming would be populateWithBrowserSpecificSettings, right? (same for populateWithAdditionalCapabilites, or something alike).

File: src/main/java/org/testeditor/fixture/web/WebDriverFixture.java [hunk 10]
```diff
@@ -288,15 +308,115 @@ private void launchFirefox() {
...
+        // set all capabilities Capabilities
+        populateCapabilities(desiredCapabilities, capabilities);
+        Object browserspecificOptions = null;
+        // set options per browser
+        switch (browserName) {
+            case BrowserType.FIREFOX:
+                FirefoxOptions firefoxOptions = new FirefoxOptions();
+                populateOptions(options, firefoxOptions);
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/fixture/web/WebDriverFixture.java]: data flow is not very clear here? which options go where? which are populated, which are read?

File: src/main/java/org/testeditor/fixture/web/WebDriverFixture.java [hunk 11]
```diff
@@ -288,15 +308,115 @@ private void launchFirefox() {
...
+                break;
+            default:
+                break;
+        }
+
+        return browserspecificOptions;
+    }
+
+    private void populateOptions(List<BrowserSetting> options, Object browserSpecificOptions) {
+        if (browserSpecificOptions instanceof FirefoxOptions) {
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/fixture/web/WebDriverFixture.java]: this called from a switch case, where the browser is known, so why have an if statement with instanceof?

File: src/main/java/org/testeditor/fixture/web/json/BrowserSettingsManager.java [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+                settings = separateBrowserSettingsForOs(elements, Platform.LINUX);
+                separateBrowserSettingsOsUnspecific(settings, elements);
+                break;
+            case MAC:
+                settings = separateBrowserSettingsForOs(elements, Platform.MAC);
+                separateBrowserSettingsOsUnspecific(settings, elements);
+                break;
+
+            default:
+                throw new RuntimeException("The current OS where the tests are executed is not supported!");
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/fixture/web/json/BrowserSettingsManager.java]: please add 'currentPlatform' to exception

File: src/main/java/org/testeditor/fixture/web/json/BrowserSettingsManager.java [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+     */
+    public static List<BrowserSetupElement> getBrowserSettings() {
+        List<BrowserSetupElement> settings = new ArrayList<>();
+        BrowserSetupReader reader = new BrowserSetupReader();
+        List<BrowserSetupElement> elements = reader.readElements("browserSetup.json");
+
+        Platform currentPlattform = Platform.getCurrent();
+        switch (currentPlattform) {
+            case VISTA:
+                settings = separateBrowserSettingsForOs(elements, Platform.WINDOWS);
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/fixture/web/json/BrowserSettingsManager.java]: code looks very uniform. seems that you only need to translate currentPlatform to some other id.

File: src/main/java/org/testeditor/fixture/web/json/BrowserSettingsManager.java [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+            "network.proxy.ssl" : "http://mysystem.proxy.server" ,
+            "network.proxy.ssl_port" : 100 
+        }
+
+     * </code>
+     */
+    public static List<BrowserSetupElement> getBrowserSettings() {
+        List<BrowserSetupElement> settings = new ArrayList<>();
+        BrowserSetupReader reader = new BrowserSetupReader();
+        List<BrowserSetupElement> elements = reader.readElements("browserSetup.json");
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/fixture/web/json/BrowserSettingsManager.java]: is "browserSetup.json" a fixed constant?

File: src/main/java/org/testeditor/fixture/web/json/BrowserSettingsManager.java [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+                break;
+
+            default:
+                throw new RuntimeException("The current OS where the tests are executed is not supported!");
+        }
+        return settings;
+
+    }
+
+    private static void separateBrowserSettingsOsUnspecific(List<BrowserSetupElement> settings,
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/fixture/web/json/BrowserSettingsManager.java]: why not keep the adding out of this method? the caller could use a returned collection with 'addAll' if wanted. You could then use java 8 stream filter to get only those elements you actually want.

File: src/test/resources/browserSetup.json [hunk 16]
```diff
@@ -1,36 +1,30 @@
 {
-    "proxy section" : {
+    "windows-ie-section" : {
+        "os" : "WINDOWS",
+        "browser" : "internet explorer",
         "capabilities" : {
-            "httpProxyUrl": "http://mysystem.proxy.server:100"
+        "ignoreProtectedModeSettings" : true ,
```
Comment (gunther-bachmann) [inline src/test/resources/browserSetup.json]: indentation?

File: src/main/java/org/testeditor/fixture/web/WebDriverFixture.java (hunk 3 above)
Comment (kbayraktar) [reply src/main/java/org/testeditor/fixture/web/WebDriverFixture.java]: No , you can add the options to an already existing DesiredCapabilities object like this

`DesiredCapabilities` capabilities = DesiredCapabilities.chrome();

// Add the WebDriver proxy capability.
Proxy proxy = new Proxy();
proxy.setHttpProxy("myhttpproxy:3337");
capabilities.setCapability("proxy", proxy);

// Add ChromeDriver-specific capabilities through ChromeOptions.
ChromeOptions options = new ChromeOptions();
options.addExtensions(new File("/path/to/extension.crx"));
capabilities.setCapability(ChromeOptions.CAPABILITY, options);
ChromeDriver driver = new ChromeDriver(capabilities);` 
[Example under Using the ChromeOptions class](https://sites.google.com/a/chromium.org/chromedriver/capabilities)

File: src/main/java/org/testeditor/fixture/web/WebDriverFixture.java (hunk 4 above)
Comment (kbayraktar) [reply src/main/java/org/testeditor/fixture/web/WebDriverFixture.java]: The internet explorer has no options :O( , so this is why there is a special treatment.

File: src/main/java/org/testeditor/fixture/web/WebDriverFixture.java (hunk 6 above)
Comment (kbayraktar) [reply src/main/java/org/testeditor/fixture/web/WebDriverFixture.java]: I agree completely to you 👍 ,  but BrowserType is a Selenium Class.

File: src/main/java/org/testeditor/fixture/web/WebDriverFixture.java (hunk 9 above)
Comment (kbayraktar) [reply src/main/java/org/testeditor/fixture/web/WebDriverFixture.java]: 👍

**Pull request merged**